model/jpa/src/main/java/org/keycloak/grant/jpa/UserGrantEntity.java

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright 2017 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.grant.jpa;
+
+import org.keycloak.models.jpa.entities.UserConsentEntity;
+
+import java.io.Serializable;
+
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.FetchType;
+import javax.persistence.Id;
+import javax.persistence.IdClass;
+import javax.persistence.JoinColumn;
+import javax.persistence.ManyToOne;
+import javax.persistence.NamedQueries;
+import javax.persistence.NamedQuery;
+import javax.persistence.Table;
+
+@NamedQueries({
+        @NamedQuery(name="userGrantByGrantIdAndUserId", query="select userGrant from UserGrantEntity userGrant where userGrant.grantId = :grantId and userGrant.userId = :userId"),
+        @NamedQuery(name="deleteUserGrantByGrantIdAndUserId", query="delete from UserGrantEntity userGrant where userGrant.grantId = :grantId and userGrant.userId = :userId"),
+})
+@Entity
+@Table(name="USER_GRANT")
+@IdClass(UserGrantEntity.Key.class)
+public class UserGrantEntity {
```

Comment:
By recent keycloak's convention, it might be better not to add or modify JPA entity because it causes a schema change. For example, Client Policies/Profiles avoid adding entities by storing JSON string itself to realm's attribute (existing key-value store).

Reply:
I see, we can stored the grants as the realm's attribute with the key "grants-extension". And the grants can be represented as follows

{
	"grants": [
	
		{
			"grantId" : ""
			"clientId" : ""
			"userId" : ""
			"authorization_details" : {}
			"scopes" : []
			"claims" : []
			"createdDate" ""
			"lastUpdatedDate" ""
		},

	]
}

WDYT?

Reply:
it's strange, realm's attribute can not get a long value:

`Caused by: org.h2.jdbc.JdbcSQLException: Value too long for column "VALUE NVARCHAR(255)": "'{""grants"":[{""grantId"":""lODVWR2zz2AfWW2CeukYdFLjvLUjKNtC41U8EYCDPsQ"",""clientId"":""client-1"",""userId"":""d1419007-7f92-43a8-9c82-4d7... (268)"; SQL statement:
update REALM_ATTRIBUTE set VALUE=? where NAME=? and REALM_ID=? [22001-197]`

Reply:
@guymoyo Could you rebase your branch on the latest keycloak master branch? It included extending Realm Attribute's size. You can found it in detail on the following :
https://github.com/keycloak/keycloak/pull/7780/files#diff-3374e8fe71e9cf4c16ff00d889e8305734333b2c059d8673aaa1060750d0ac79

Reply:
Thanks, I looked at this PR. It's clear however if I rebase on the latest keycloak master branch, should i not have a lot of conflicts / commits when trying to merge on tnorimat / keycloak master branch because it is not yet up to date?

Reply:
@guymoyo How about the following?

1. I will rebase tnorimat/master on keycloak/master
2. you will rebase your branch on keycloak/master
3. you will push -f your rebased branch to tnorimat

But I'm not sure it works well...